File: demo-python-service/patterns/deployment/canary.py
"""
Canary Deployment Pattern Implementation
Traffic splitting with metrics and rollback capability
"""
import hashlib
import time
from typing import Dict, Any, Optional
from dataclasses import dataclass

@dataclass
class CanaryMetrics:
    canary_requests: int = 0
    stable_requests: int = 0
    canary_errors: int = 0
    stable_errors: int = 0
    canary_response_times: list = None
    stable_response_times: list = None
    
    def __post_init__(self):
        if self.canary_response_times is None:
            self.canary_response_times = []
        if self.stable_response_times is None:
            self.stable_response_times = []

class CanaryDeployment:
    def __init__(self, canary_percentage: int = 10, max_error_rate: float = 0.05):
        self.canary_percentage = canary_percentage
        self.max_error_rate = max_error_rate
        self.metrics = CanaryMetrics()
        self.enabled = True
        self.auto_rollback = True
        self.rollback_threshold = 10  # Number of errors before rollback
    
    def should_use_canary(self, context: Optional[Dict[str, Any]] = None) -> bool:
        """Determine if request should use canary version"""
        if not self.enabled:
            return False
        
        # Check if we should rollback due to high error rate
        if self.auto_rollback and self._should_rollback():
            self.enabled = False
            return False
        
        if context and context.get('user_id'):
            # Consistent user-based routing
            user_id = str(context['user_id'])
            hash_value = int(hashlib.md5(f"canary:{user_id}".encode()).hexdigest(), 16) % 100
            return hash_value < self.canary_percentage
        
        # Random routing if no user context
        import random
        return random.randint(0, 99) < self.canary_percentage
    
    def process_request(self, data: Dict[str, Any], context: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Process request with canary deployment logic"""
        start_time = time.time()
        use_canary = self.should_use_canary(context)
        
        try:
            if use_canary:
                result = self._process_canary_version(data)
                self.metrics.canary_requests += 1
                response_time = time.time() - start_time
                self.metrics.canary_response_times.append(response_time)
                
                return {
                    **result,
                    'version': 'v2-canary',
                    'canary': True,
                    'response_time': response_time
                }
            else:
                result = self._process_stable_version(data)
                self.metrics.stable_requests += 1
                response_time = time.time() - start_time
                self.metrics.stable_response_times.append(response_time)
                
                return {
                    **result,
                    'version': 'v1-stable',
                    'canary': False,
                    'response_time': response_time
                }
        
        except Exception as e:
            if use_canary:
                self.metrics.canary_errors += 1
            else:
                self.metrics.stable_errors += 1
            
            # Return error response
            return {
                'error': str(e),
                'version': 'v2-canary' if use_canary else 'v1-stable',
                'canary': use_canary,
                'response_time': time.time() - start_time
            }
    
    def _process_canary_version(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Process with canary (new) version"""
        value = data.get('value', 0)
        
        # Enhanced algorithm with additional features
        result = value * 3 + 1  # New calculation
        
        return {
            'result': result,
            'algorithm': 'enhanced-v2',
            'features': ['optimization', 'caching', 'validation']
        }
    
    def _process_stable_version(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Process with stable (current) version"""
        value = data.get('value', 0)
        
        # Standard algorithm
        result = value * 2  # Current calculation
        
        return {
            'result': result,
            'algorithm': 'standard-v1',
            'features': ['basic']
        }
    
    def _should_rollback(self) -> bool:
        """Check if canary should be rolled back due to high error rate"""
        if self.metrics.canary_requests < 10:  # Need minimum requests
            return False
        
        canary_error_rate = self.metrics.canary_errors / self.metrics.canary_requests
        stable_error_rate = (self.metrics.stable_errors / self.metrics.stable_requests 
                           if self.metrics.stable_requests > 0 else 0)
        
        # Rollback if canary error rate is significantly higher
        return (canary_error_rate > self.max_error_rate or 
                canary_error_rate > stable_error_rate * 2)
    
    def get_metrics(self) -> Dict[str, Any]:
        """Get deployment metrics"""
        total_requests = self.metrics.canary_requests + self.metrics.stable_requests
        total_errors = self.metrics.canary_errors + self.metrics.stable_errors
        
        canary_error_rate = (self.metrics.canary_errors / self.metrics.canary_requests 
                           if self.metrics.canary_requests > 0 else 0)
        stable_error_rate = (self.metrics.stable_errors / self.metrics.stable_requests 
                           if self.metrics.stable_requests > 0 else 0)
        
        # Calculate average response times
        avg_canary_time = (sum(self.metrics.canary_response_times) / len(self.metrics.canary_response_times)
                          if self.metrics.canary_response_times else 0)
        avg_stable_time = (sum(self.metrics.stable_response_times) / len(self.metrics.stable_response_times)
                          if self.metrics.stable_response_times else 0)
        
        return {
            'canary_enabled': self.enabled,
            'canary_percentage': self.canary_percentage,
            'total_requests': total_requests,
            'canary_requests': self.metrics.canary_requests,
            'stable_requests': self.metrics.stable_requests,
            'canary_errors': self.metrics.canary_errors,
            'stable_errors': self.metrics.stable_errors,
            'canary_error_rate': canary_error_rate,
            'stable_error_rate': stable_error_rate,
            'avg_canary_response_time': avg_canary_time,
            'avg_stable_response_time': avg_stable_time,
            'should_rollback': self._should_rollback()
        }
    
    def set_canary_percentage(self, percentage: int):
        """Update canary traffic percentage"""
        self.canary_percentage = max(0, min(100, percentage))
    
    def enable_canary(self):
        """Enable canary deployment"""
        self.enabled = True
    
    def disable_canary(self):
        """Disable canary deployment (rollback)"""
        self.enabled = False
    
    def reset_metrics(self):
        """Reset all metrics"""
        self.metrics = CanaryMetrics()

File: demo-node-service/patterns/deployment/canary.js
/**
 * Canary Deployment Pattern Implementation
 * Traffic splitting with metrics and rollback capability
 */
const crypto = require('crypto');

class CanaryMetrics {
    constructor() {
        this.canaryRequests = 0;
        this.stableRequests = 0;
        this.canaryErrors = 0;
        this.stableErrors = 0;
        this.canaryResponseTimes = [];
        this.stableResponseTimes = [];
    }

    reset() {
        this.canaryRequests = 0;
        this.stableRequests = 0;
        this.canaryErrors = 0;
        this.stableErrors = 0;
        this.canaryResponseTimes = [];
        this.stableResponseTimes = [];
    }
}

class CanaryDeployment {
    constructor(canaryPercentage = 10, maxErrorRate = 0.05) {
        this.canaryPercentage = canaryPercentage;
        this.maxErrorRate = maxErrorRate;
        this.metrics = new CanaryMetrics();
        this.enabled = true;
        this.autoRollback = true;
        this.rollbackThreshold = 10; // Number of errors before rollback
    }

    shouldUseCanary(context = {}) {
        if (!this.enabled) {
            return false;
        }

        // Check if we should rollback due to high error rate
        if (this.autoRollback && this._shouldRollback()) {
            this.enabled = false;
            console.log('Canary deployment disabled due to high error rate');
            return false;
        }

        if (context.userId) {
            // Consistent user-based routing
            const hash = crypto.createHash('md5')
                .update(`canary:${context.userId}`)
                .digest('hex');
            const hashValue = parseInt(hash.substring(0, 8), 16) % 100;
            return hashValue < this.canaryPercentage;
        }

        // Random routing if no user context
        return Math.floor(Math.random() * 100) < this.canaryPercentage;
    }

    async processRequest(data, context = {}) {
        const startTime = Date.now();
        const useCanary = this.shouldUseCanary(context);

        try {
            let result;
            if (useCanary) {
                result = await this._processCanaryVersion(data);
                this.metrics.canaryRequests++;
                const responseTime = Date.now() - startTime;
                this.metrics.canaryResponseTimes.push(responseTime);

                return {
                    ...result,
                    version: 'v2-canary',
                    canary: true,
                    responseTime
                };
            } else {
                result = await this._processStableVersion(data);
                this.metrics.stableRequests++;
                const responseTime = Date.now() - startTime;
                this.metrics.stableResponseTimes.push(responseTime);

                return {
                    ...result,
                    version: 'v1-stable',
                    canary: false,
                    responseTime
                };
            }
        } catch (error) {
            if (useCanary) {
                this.metrics.canaryErrors++;
            } else {
                this.metrics.stableErrors++;
            }

            // Return error response
            return {
                error: error.message,
                version: useCanary ? 'v2-canary' : 'v1-stable',
                canary: useCanary,
                responseTime: Date.now() - startTime
            };
        }
    }

    async _processCanaryVersion(data) {
        // Enhanced algorithm with additional features
        const value = data.value || 0;
        
        // Simulate some processing time
        await this._sleep(Math.random() * 50);
        
        // New calculation with enhancements
        const result = value * 3 + 1;

        return {
            result,
            algorithm: 'enhanced-v2',
            features: ['optimization', 'caching', 'validation'],
            enhancement: 'improved-performance'
        };
    }

    async _processStableVersion(data) {
        // Standard algorithm
        const value = data.value || 0;
        
        // Simulate some processing time
        await this._sleep(Math.random() * 30);
        
        // Current calculation
        const result = value * 2;

        return {
            result,
            algorithm: 'standard-v1',
            features: ['basic']
        };
    }

    _shouldRollback() {
        if (this.metrics.canaryRequests < 10) { // Need minimum requests
            return false;
        }

        const canaryErrorRate = this.metrics.canaryErrors / this.metrics.canaryRequests;
        const stableErrorRate = this.metrics.stableRequests > 0 
            ? this.metrics.stableErrors / this.metrics.stableRequests 
            : 0;

        // Rollback if canary error rate is significantly higher
        return canaryErrorRate > this.maxErrorRate || 
               canaryErrorRate > stableErrorRate * 2;
    }

    getMetrics() {
        const totalRequests = this.metrics.canaryRequests + this.metrics.stableRequests;
        const totalErrors = this.metrics.canaryErrors + this.metrics.stableErrors;

        const canaryErrorRate = this.metrics.canaryRequests > 0 
            ? this.metrics.canaryErrors / this.metrics.canaryRequests 
            : 0;
        const stableErrorRate = this.metrics.stableRequests > 0 
            ? this.metrics.stableErrors / this.metrics.stableRequests 
            : 0;

        // Calculate average response times
        const avgCanaryTime = this.metrics.canaryResponseTimes.length > 0
            ? this.metrics.canaryResponseTimes.reduce((a, b) => a + b, 0) / this.metrics.canaryResponseTimes.length
            : 0;
        const avgStableTime = this.metrics.stableResponseTimes.length > 0
            ? this.metrics.stableResponseTimes.reduce((a, b) => a + b, 0) / this.metrics.stableResponseTimes.length
            : 0;

        return {
            canaryEnabled: this.enabled,
            canaryPercentage: this.canaryPercentage,
            totalRequests,
            canaryRequests: this.metrics.canaryRequests,
            stableRequests: this.metrics.stableRequests,
            canaryErrors: this.metrics.canaryErrors,
            stableErrors: this.metrics.stableErrors,
            canaryErrorRate,
            stableErrorRate,
            avgCanaryResponseTime: avgCanaryTime,
            avgStableResponseTime: avgStableTime,
            shouldRollback: this._shouldRollback(),
            timestamp: Date.now()
        };
    }

    setCanaryPercentage(percentage) {
        this.canaryPercentage = Math.max(0, Math.min(100, percentage));
    }

    enableCanary() {
        this.enabled = true;
        console.log('Canary deployment enabled');
    }

    disableCanary() {
        this.enabled = false;
        console.log('Canary deployment disabled (rollback)');
    }

    resetMetrics() {
        this.metrics.reset();
        console.log('Canary metrics reset');
    }

    // Health check for canary deployment
    getHealthStatus() {
        const metrics = this.getMetrics();
        const isHealthy = !metrics.shouldRollback && metrics.canaryErrorRate <= this.maxErrorRate;

        return {
            healthy: isHealthy,
            status: this.enabled ? 'ACTIVE' : 'DISABLED',
            canaryPercentage: this.canaryPercentage,
            errorRate: metrics.canaryErrorRate,
            maxErrorRate: this.maxErrorRate,
            autoRollback: this.autoRollback,
            metrics: metrics
        };
    }

    _sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// Middleware for Express.js
function canaryMiddleware(canaryDeployment) {
    return async (req, res, next) => {
        req.canaryDeployment = canaryDeployment;
        req.useCanary = canaryDeployment.shouldUseCanary({
            userId: req.headers['x-user-id'] || req.ip,
            requestId: req.headers['x-request-id']
        });
        next();
    };
}

module.exports = {
    CanaryDeployment,
    CanaryMetrics,
    canaryMiddleware
};